Modules/_winapi.c

```diff
@@ -774,12 +774,139 @@ gethandle(PyObject* obj, const char* name)
+    }
+    kwnames = PyTuple_New(1);
+    if (kwnames == NULL) {
+        goto error;
+    }
+    assert(PyTuple_SetItem(kwnames, 0, PyUnicode_FromString("key")) == 0);
+    PyObject *args[] = { keys, keyfunc };
+    if (PyObject_VectorcallMethod(&_Py_ID(sort), args, 1, kwnames) == NULL) {
+        goto error;
+    }
+
+    result = PyDict_New();
+    if (result == NULL) {
+        goto error;
+    }
+    wchar_t *prev_key_string = NULL;
+
+    for (int i=0; i<PyList_GET_SIZE(keys); i++) {
+        PyObject *key = PyList_GET_ITEM(keys, i);
+        if (key == NULL) {
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
+        }
+        PyObject* value = PyObject_GetItem(environment, key);
+        if (value == NULL) {
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
+        }
+
+        if (! PyUnicode_Check(key) || ! PyUnicode_Check(value)) {
+            PyErr_SetString(PyExc_TypeError,
+                "environment can only contain strings");
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
+        }
+        if (PyUnicode_FindChar(key, '\0', 0, PyUnicode_GET_LENGTH(key), 1) != -1 ||
+            PyUnicode_FindChar(value, '\0', 0, PyUnicode_GET_LENGTH(value), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "embedded null character");
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
+        }
+        /* Search from index 1 because on Windows starting '=' is allowed for
+           defining hidden environment variables. */
+        if (PyUnicode_GET_LENGTH(key) == 0 ||
+            PyUnicode_FindChar(key, '=', 1, PyUnicode_GET_LENGTH(key), 1) != -1)
+        {
+            PyErr_SetString(PyExc_ValueError, "illegal environment variable name");
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
+        }
+
+        if (i == 0) {
+            prev_key_string = PyUnicode_AsWideCharString(key, NULL);
+            if (prev_key_string == NULL) {
+                Py_XDECREF(result);
+                result = NULL;
+                goto error;
+            }
+            if (PyObject_SetItem(result, key, value) < 0) {
+                Py_XDECREF(result);
+                result = NULL;
+                goto error;
+            }
+            continue;
+        }
+
+        wchar_t *key_string = PyUnicode_AsWideCharString(key, NULL);
+        if (key_string == NULL) {
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
+        }
+        if (CompareStringOrdinal(prev_key_string, -1, key_string, -1, TRUE) == CSTR_EQUAL) {
+            continue;
+        }
+        if (PyObject_SetItem(result, key, value) < 0) {
+            Py_XDECREF(result);
+            result = NULL;
+            goto error;
+        }
+        PyMem_Free(prev_key_string);
+        prev_key_string = key_string;
+    }
+
+error:
```

Comment:
```suggestion
    goto cleanup;
error:
    Py_XDECREF(result);
    result = NULL;
cleanup:
```

This will reduce a significant number of lines from above, which I'll mark with suggestions.